File: Java/08_Greedy Algorithm/M-GFG-Minimum Coins to Make Sum.java
// GFG : Find Minimum Number Of Coins That Make A Change.
// Coin Change - Minimum Coins to Make Sum
// https://www.geeksforgeeks.org/problems/number-of-coins1824/1

import java.util.Arrays;

class Solution {

    public int minCoins(int[] coins, int sum) {
        int[] dp = new int[sum + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;
        for (int i = 1; i <= sum; i++) {
            for (int coin : coins) {
                if (i - coin >= 0 && dp[i - coin] != Integer.MAX_VALUE) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[sum] == Integer.MAX_VALUE ? -1 : dp[sum];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] coins = { 1, 2, 5 };
        int sum = 11;
        System.out.println(solution.minCoins(coins, sum));
    }
}

File: JavaScript/08_Greedy Algorithm/M-GFG-Minimum Coins to Make Sum.js
// GFG: Find Minimum Number Of Coins That Make A Change
// Coin Change - Minimum Coins to Make Sum
// https://www.geeksforgeeks.org/problems/number-of-coins1824/1

/**
 * @param {number[]} coins
 * @param {number} sum
 * @return {number}
 */
function minCoins(coins, sum) {
    const dp = new Array(sum + 1).fill(Infinity);
    dp[0] = 0;

    for (let i = 1; i <= sum; i++) {
        for (const coin of coins) {
            if (i - coin >= 0 && dp[i - coin] !== Infinity) {
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }
    }

    return dp[sum] === Infinity ? -1 : dp[sum];
}

const coins = [1, 2, 5];
const sum = 11;
console.log(minCoins(coins, sum)); // Output: 3
